### PR TITLE
puppetserver AIO uses a different vardir

### PR DESCRIPTION
**Pull request by mmckinst**
puppetserver runs as the puppet user, the puppet agent normally runs as
the root user. When the two try to share the same vardir,
puppetserver (running as the puppet user) can't always read/write/create
directories if the puppet agent creates them as the root user. The
problem is described more thoroughly in
https://tickets.puppetlabs.com/browse/SERVER-357 .

### CODE REVIEW

File: manifests/params.pp [hunk 1]
```diff
@@ -23,15 +23,17 @@
   $show_diff           = false
   $module_repository   = undef
   if versioncmp($::puppetversion, '4.0') < 0 {
-    $configtimeout           = 120
-    $server_puppetserver_dir = '/etc/puppetserver'
-    $server_ruby_load_paths  = []
-    $server_jruby_gem_home   = '/var/lib/puppet/jruby-gems'
+    $configtimeout              = 120
+    $server_puppetserver_dir    = '/etc/puppetserver'
+    $server_puppetserver_vardir = undef
```
Comment (ekohl) [inline manifests/params.pp]: I think we also support puppetserver 1 which is puppet 3. Won't this break and should it be ```$vardir```?

File: manifests/init.pp [hunk 2]
```diff
@@ -527,6 +527,10 @@
 #                                     disable in case CA is delegated to a separate instance
 #                                     type:boolean
 #
+# $server_puppetserver_vardir::       The path of the puppetserver var dir
+#                                     Defaults to '/opt/puppetlabs/server/data/puppetserver'
```
Comment (ekohl) [inline manifests/init.pp]: Given we have multiple defaults depending on the version, I think we should leave out this line

**Comment by mmckinst** [comment]
@ekohl I think 5519203 should work in terms of puppetserver 1.

**Comment by domcleal** [comment]
It looks good generally, but it doesn't work with Foreman's ENC as the facts YAML files now move over to the master vardir, while foreman::puppetmaster is configured to use the agent vardir.

I think updating foreman::puppetmaster's puppet_home parameter in puppet::server::config to point to the master dir should fix it.

File: spec/classes/puppet_server_config_spec.rb [hunk 3]
```diff
@@ -15,29 +15,31 @@
...
+        vardir              = '/opt/puppetlabs/puppet/cache'
+        puppetserver_vardir = '/opt/puppetlabs/server/data/puppetserver'
+        ssldir              = '/etc/puppetlabs/puppet/ssl'
+        sharedir            = '/opt/puppetlabs/puppet'
+        etcdir              = '/etc/puppetlabs/puppet'
+        puppetcacmd         = '/opt/puppetlabs/bin/puppet cert'
+        additional_facts    = {:rubysitedir => '/opt/puppetlabs/puppet/lib/ruby/site_ruby/2.1.0'}
       end
 
       if os_facts[:osfamily] == 'FreeBSD'
```
Comment (domcleal) [inline spec/classes/puppet_server_config_spec.rb]: I think you need to set `puppetserver_vardir` here too, as there are some test failures:

<pre>
  1) puppet::server::config on freebsd-10-amd64 with no custom parameters should set up the ENC

     Failure/Error: })

       expected that the catalogue would contain Class[foreman::puppetmaster] with puppet_home set to "/var/lib/puppet" but it is set to "/var/puppet"
</pre>

File: manifests/server/config.pp [hunk 4]
```diff
@@ -192,7 +192,7 @@
     class {'::foreman::puppetmaster':
       foreman_url    => $::puppet::server::foreman_url,
       receive_facts  => $::puppet::server::server_facts,
-      puppet_home    => $puppet::vardir,
+      puppet_home    => $::puppet::server::server_puppetserver_vardir,
```
Comment (ekohl) [inline manifests/server/config.pp]: Shouldn't this be ```$::puppet::server::puppetserver_vardir```?

**Comment by mmoll** [comment]
I opended mmckinst/puppet-puppet#1 with fixes to the tests.

**Comment by mmckinst** [comment]
@mmoll thanks so much for taking care of that!

File: manifests/server/puppetserver.pp [hunk 5]
```diff
@@ -25,6 +25,9 @@
 # * `server_puppetserver_dir`
 # Puppetserver config directory
 #
+# * `server_puppetserver_vardir`
+# Puppetserver config directory
```
Comment (mmoll) [inline manifests/server/puppetserver.pp]: that comment needs updating

**Comment by mmoll** [comment]
One small nitpick, :+1: otherwise.
Tested and found working.

**Comment by mmoll** [comment]
:+1:

**Comment by ekohl** [comment]
Thanks! I took the liberty of squash all commits, with @mmoll's permission.

**Pull request closed without merging**